cmake/compiler_config.cmake

```diff
@@ -23,6 +23,8 @@ set(MSVC_FLAGS
     # we want to be as strict as possible
     /W3
     $<$<BOOL:${BUILD_STRICT_MODE}>:/WX>
+    # enable two-phase name lookup and other strict checks (binding a non-const reference to a temporary, etc..)
+    $<$<BOOL:$<VERSION_GREATER:${PXR_VERSION},2002>>:/permissive->
```

Comment:
By checking the USD version here, you make assumptions that nobody changes the default boost version used to build the USD library. I wouldn't make this assumption and instead, add a check that validates the boost version since this is the culprit to the problem.

Reply:
Not quite sure if I follow what you mean by "changing the default boost version". 

We can't and shouldn't rely on Boost version at maya-usd top level cmake.

`${Boost_VERSION} ` is not populated unless you call `find_package(Boost REQUIRED)`. We tried so hard to eliminate such calls and bringing it back again doesn't make much sense.

Second, it makes more sense to rely on USD version because moving forward with USD 20.05, Pixar uses boost-1_70 and it would be highly unlikely that they decide to switch to an old version of boost for any future releases. 

https://github.com/PixarAnimationStudios/USD/blob/ebac0a8b6703f4fa1c27115f1f013bb9819662f4/build_scripts/build_usd.py#L631

I agree with you that the culprit is boost. If it helps I can add a comment to clarify it.

Reply:
Not everyone builds USD with the default configuration and you exactly pointed to the place where one can change the boost version. This means you can build older versions of USD with a newer boost version (or vice versa).

Reply:
I see and understand both your points. The only component that relies on Boost in the repo is "plugin/al" (that is the only place "find_package(Boost)" is called, thus here in this top-level compiler_config.cmake we don't know the boost version. I agree with Hamed that we should not add a find_package(Boost) to the top-level cmake. So the only thing we can rely on here is the USD version. Unless you have another idea Krystian?

Reply:
The comment in 400532bb1b0531cf83c9d556a2bdffd942d6f6fb talks about boost python - we do use boost python for core module python bindings. What am I missing that prevents us from getting the boost version?

Reply:
It is a bit confusing. Yes we use boost python in the core, which we get via inheritance from linking with USD libs such as Tf. From USD we get the include path to boost (inside the USD artifact) and the boost lib(s). Using boost in this fashion does **not** give us the boost version. 

One way to get the boost version would be to read it from the boost file _version.hpp_. I suppose we modify our FindUSD.cmake to do this just like it gets the USD_VERSION from prx.h.

Reply:
@kxl-adsk The boost python dependency is propagated as soon as we link against ANY of the Pixar USD libraries ( e.g usd, usdImaging, etc... ).

![image](https://user-images.githubusercontent.com/48299423/111803920-eb15fc00-88a5-11eb-8214-2d9c985b0b7e.png)

If you recall, such dependencies are leaking from Pixar libraries. So even if you have a target that has nothing to do with Boost or python, you still need to make sure that target links against such libraries.

Reply:
I see, this makes much more sense now. Thank you. I'm guessing what Sean proposed is still pretty much doable?

> modify our FindUSD.cmake to do this just like it gets the USD_VERSION from prx.h

Reply:
I would prefer to not make this change for the sake of this PR. We could revisit this at any time **if** any problems rises as a result of this change.

Reply:
The change is FindUSD.cmake looks straight forward and we already have a pattern to copy from. I spoke with @seando-adsk and he will help with it.